### PR TITLE
bug: PhpdocSeparationFixer - Make groups handling more flexible

### LINKED ISSUE #6635: PhpdocSeparationFixer started grouping "other" annotations together
After merging #6501 PhpdocSeparationFixer started removing blank lines between annotations similar to: `@Method('GET')`, `@OA\Property(...)`, `@Assert\NotBlank`. Before, they could be freely grouped or separated and the fixer did not care.

## Configuration
```php
'phpdoc_separation' => true,
```

## Code example
```php
<?php declare(strict_types=1);

use OpenApi\Annotations as OA;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method;
use Symfony\Component\Validator\Constraints as Assert;

class Test
{
    /**
     * @Assert\NotBlank
     * @Assert\Email
     *
     * @OA\Property(type="string")
     *
     * @Method('GET')
     */
    public string $property;
}
```

## Fixer
Before:
```diff
PHP CS Fixer 3.10.0 (76d7da6) Oliva by Fabien Potencier and Dariusz Ruminski.
PHP runtime: 8.1.10
Loaded config default from "/private/tmp/-/.php-cs-fixer.dist.php".
Using cache file ".php-cs-fixer.cache".
Paths from configuration file have been overridden by paths provided as command arguments.
.                                                                                                                                                                                             1 / 1 (100%)
Legend: .-no changes, F-fixed, S-skipped (cached or empty file), I-invalid file syntax (file ignored), E-error

Checked all files in 0.003 seconds, 14.547 MB memory used
```

After:
```diff
PHP CS Fixer 3.11.0 Oliva by Fabien Potencier and Dariusz Ruminski.
PHP runtime: 8.1.10
Loaded config default from "/private/tmp/-/.php-cs-fixer.dist.php".
Using cache file ".php-cs-fixer.cache".
Paths from configuration file have been overridden by paths provided as command arguments.
F                                                                                                                                                                                             1 / 1 (100%)
Legend: .-no changes, F-fixed, S-skipped (cached or empty file), I-invalid file syntax (file ignored), E-error
   1) Test.php (phpdoc_separation)
      ---------- begin diff ----------
--- /private/tmp/-/Test.php
+++ /private/tmp/-/Test.php
@@ -9,10 +9,8 @@
     /**
      * @Assert\NotBlank
      * @Assert\Email
-     *
      * @OA\Property(type="string")
-     *
      * @Method('GET')
      */
     public string $property;
 }
\ No newline at end of file

      ----------- end diff -----------


Checked all files in 0.008 seconds, 18.000 MB memory used
```

## It's not a bug, it's a feature?
I'm not sure what is the correct behavior here. The rule's description states:
> Annotations of a different type are separated by a single blank line.

If this was true and given that: `@OA\Property` and `@Assert\NotBlank` are of different type I would expect the fixer to convert
```php
/**
 * @OA\Property(type="string", format="email")
 * @Assert\NotBlank()
 */
public string $foo;
```
into
```php
/**
 * @OA\Property(type="string", format="email")
 * 
 * @Assert\NotBlank()
 */
public string $foo;
```
Neither v3.10.0 nor v3.11.0 do it. It's because all tags not matching [certain regex](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/abf6c314b99bd6dc156ab8e39967956141d8594c/src/DocBlock/Tag.php#L63) are being treated as "other". For this exact reason v3.11.0 forces them to be grouped together. I'm not sure why v3.10 does not.

Personally I liked it better when fixer ignored custom annotations while still forcing separation of standard tags, but I know it's a matter of preference.

## Configurability
It's obvious that grouping annotations is non-trivial so #6501 is a great feature to have. But now it's impossible to keep the rule turned on while still being allowed to have some custom annotations separated. After upgrading to v3.11 we were forced to turn off the rule because it reduced the readability of our phpdocs.

I don't know what is the way to go here. I think I would expect the fixer to convert
```php
// example1

/**
 * @param non-empty-string
 * @throws \LogicException
 * @return non-empty-string
 *
 * @Foo\Bar
 * @Foo\Baz
 *
 * @SomeOtherTag()
 */
public function property(string $foo): string
```

or

```php
// example2

/**
 * @param non-empty-string
 * @throws \LogicException
 * @return non-empty-string
 *
 * @Foo\Bar
 * @Foo\Baz
 * @SomeOtherTag()
 */
public function property(string $foo): string
```

into 

```php
// expectedResult

/**
 * @param non-empty-string
 * 
 * @throws \LogicException
 * 
 * @return non-empty-string
 *
 * @Foo\Bar
 * @Foo\Baz
 * 
 * @SomeOtherTag()
 */
public function property(string $foo): string
```

_Note that conversion example1 -> expectedResult is possible in v3.10.0 but not in v3.11.0_

The configuration could look like this:
```php
'phpdoc_separation' => [
    'groups' => [
        ['deprecated', 'link', 'see', 'since'],
        ['author', 'copyright', 'license'],
        ['category', 'package', 'subpackage'],
        ['property', 'property-read', 'property-write'],
        ['foo'],
        ['someothertag'],
    ]
],
```

### PR DESCRIPTION
**Pull request by julienfalque**
Fixes #6635.

This is a proposal to change the way the option works a bit.

Before, two tags were grouped together if they were both explicity listed in the same group and separated otherwise.

Now, they are still grouped if they are both in the same group, but they are separated *only* if one is in a group and the other is in another group or not in a group at all. If both are not in any group they are left untouched, meaning they remain grouped or separated.

So if we want to separate a single tag from others, we now have to put it explicity in a group too. This is kind of a BC break though, but I think it's reasonable given the fact that the current behavior is a bit unexpected.

This PR also adds support for more complex tags such as Doctrine annotations; and deprecates `TagComparator` which implies a minor release if merged.

### CODE REVIEW

File: tests/AutoReview/ProjectCodeTest.php [hunk 1]
```diff
@@ -526,34 +525,6 @@ public function testAllCodeContainSingleClassy(string $className): void
         static::assertNull($tokens->getNextNonWhitespace($classyEndIndex), sprintf('File "%s" should only contains a single classy.', $file));
     }
 
-    /**
-     * @dataProvider provideSrcClassCases
-     */
-    public function testThereIsNoTriggerErrorUsedDirectly(string $className): void
```
Comment (julienfalque) [inline tests/AutoReview/ProjectCodeTest.php:532]: The added deprecation in this PR is not intended for end-users but for library authors. I don't know how to keep this test and tell it which `trigger_error()` calls to allow reliably.

**Review by kubawerlos — APPROVED** [review]

File: src/DocBlock/TagComparator.php [hunk 2]
```diff
@@ -42,6 +44,8 @@ final class TagComparator
      */
     public static function shouldBeTogether(Tag $first, Tag $second, array $groups = self::DEFAULT_GROUPS): bool
     {
+        @trigger_error('Method '.__METHOD__.' is deprecated and will be removed in version 4.0.');
```
Comment (julienfalque) [inline src/DocBlock/TagComparator.php:47]: This should use `E_DEPRECATED`. Also test missing.

**Comment by VincentLanglet** [comment]
> So if we want to separate a single tag from others, we now have to put it explicity in a group too. This is kind of a BC break though

This would be an important and annoying BC-break for static-analysis user which phpdoc like `phpstan-param`, `psalm-param`, `phpstan-return`, `phpstan-template` ...

Adding all the existing annotation to the config is way too complicated.

> but I think it's reasonable given the fact that the current behavior is a bit unexpected.

The description of the rule is 
```
Annotations in PHPDoc should be grouped together so that annotations of the same type immediately follow each other. Annotations of a different type are separated by a single blank line.
```

So it's expected. "Annotations of a different type are separated by a single blank line." is meaning that any different annotation are separated by one line, there was nothing explaining it was only the annotation from the groups.

IMHO, there are multiple things in the link issue
```
@Assert\NotBlank
@Assert\Email

@OA\Property(type="string")

@Method('GET')
```
First, it would be interesting to understand why the spaces are remove between those tags ; which are the opposite behavior of what is expected. I would expect instead instead
```
@Assert\NotBlank

@Assert\Email

@OA\Property(type="string")

@Method('GET')
```
Then, since people are generally grouping the Assert\* together, we should provide an easy way to do this (adding all the Assert\* to the config is too complicated). I see two solutions
- Considering that the tag content is only the part before the first `\`, so we would have this behavior by default
- Adding support for regex in the config, so it would be possible to add `['Assert\*']` to group all the asserts.

**Comment by julienfalque** [comment]
@VincentLanglet You're right, I reworked the changes to add a `skip_unlisted_annotations` option. The behavior I described above only applies when the option is enabled. I still think this should be the default behavior so I propose to set the default value of this option to `true` in next major release.

> First, it would be interesting to understand why the spaces are remove between those tags ; which are the opposite behavior of what is expected

This was because `\` was not allowed in the tag names. And unreadble tags were always considered to be in the same group. Now `\` is allowed and in case a tag still cannot be read, it should be skipped.

Finally, I added support for `*` wildcard in tag names.

All those changes should give enough flexibility to tweak the rule to one's needs but it will likely require updating the configuration when updating the tool to avoid adding lines between tags such as `@ORM\…` or `@Assert\…`.

**Comment by VincentLanglet** [comment]
> @VincentLanglet You're right, I reworked the changes to add a `skip_unlisted_annotations` option. The behavior I described above only applies when the option is enabled. I still think this should be the default behavior so I propose to set the default value of this option to `true` in next major release.
> 
> > First, it would be interesting to understand why the spaces are remove between those tags ; which are the opposite behavior of what is expected
> 
> This was because `\` was not allowed in the tag names. And unreadble tags were always considered to be in the same group. Now `\` is allowed and in case a tag still cannot be read, it should be skipped.
> 
> Finally, I added support for `*` wildcard in tag names.
> 
> All those changes should give enough flexibility to tweak the rule to one's needs but it will likely require updating the configuration when updating the tool to avoid adding lines between tags such as `@ORM\…` or `@Assert\…`.

Awesome changes. This seems to be a perfect solution.

**Comment by keradus** [comment]
Big thanks!

**Pull request merged**